Add security guide

adellape opened the pull request:
Trello: https://trello.com/c/yRF0aKkX

Preview build:

http://file.rdu.redhat.com/~adellape/032217/security_guide/security/

adellape commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Security Guide](https://trello.com/c/yRF0aKkX/237-security-guide)

php-coder commented on security/deployment.adoc:
```diff
@@ -0,0 +1,85 @@
+- Capabilities a container can request to be added.
+- Use of host directories as volumes.
+- SELinux context of the container.
+- Container user ID.
```
Readers could think that this is a full list of possible aspects that can be managed by SCC but in fact it's only part of this list.

adellape replied:
Thanks, I'll make it more clear it's a partial list and that more are possible / direct them to the Further Reading section for more.

php-coder commented:
Image `security/images/bringing_it_all_together.png` has a typo in "contaner" word. Also "kubernetes" word is highlighted like unknown word. Perhaps, this image could be fixed.

adellape commented:
@openshift/team-documentation Feel free to peer review this in its current state.

ahardin-rh commented on security/authn_authz.adoc:
```diff
@@ -0,0 +1,67 @@
+- _{product-title} Developer Guide_
+** xref:{sourcedir}dev_guide/authentication.adoc#web-console-authentication[Web Console Authentication]
+** xref:{sourcedir}dev_guide/authentication.adoc#cli-authentication[CLI Authentication]
+- xref:{sourcedir}cli_reference/index.adoc#cli-reference-index[_{product-title} CLI Reference_]
```
oooOoo I like the {sourcedir} format :smiley_cat:

adellape replied:
Oh yeah, I was trying that out with this PR. Need to bring it up as a discussion point/idea in next team mtg.

bfallonf replied:
I saw this. I'm not too sure what the benefits would be, and I think this would be well worth discussing before implementing here. I don't want another case of "one of us starts doing a thing and before Brice know it, it's a standard".

adellape replied:
@bfallonf Nah, I was just testing out the idea in this PR, since it's self-contained within this new guide. Been planning to bring it up as a discussion point in a future meeting after I tried it out more.

ahardin-rh commented on security/container_content.adoc:
```diff
@@ -0,0 +1,53 @@
+toc::[]
+
+[[security-content-whats-in-your-container]]
+== What's In Your Container?
```
Title case would be: What's in Your Container?

ahardin-rh commented on security/hosts_multitenancy.adoc:
```diff
@@ -0,0 +1,105 @@
+certification of the Linux Container Framework Support.
+
+[[security-host-multi-tenancy]]
+== Multi-tenancy: Virtualization vs Containers
```
Versus (per IBM Style Guide)

ahardin-rh commented on security/index.adoc:
```diff
@@ -0,0 +1,85 @@
+// image::whatarecontainers.png["What Are Containers?", align="center"]
+
+[options="header",frame="all",grid="none"]
+|===
```
I :heart: this table versus the original graphic.

vikram-redhat commented:
Looks great @adellape ! Can we get someone from QE to eyeball it as well?

@aheslin @mburke5678 @bfallonf @bmcelvee - please review.

bfallonf commented on _topic_map.yml:
```diff
@@ -214,6 +214,38 @@ Topics:
     Distros: openshift-enterprise,openshift-dedicated
 
 ---
+Name: Complete Container Security
```
I'm wondering if this should just be "Container Security". I'm thinking that having "Complete" would come under the same umbrella of why we don't use "best practices", ie, we can't claim that it's the _best_, much like we shouldn't claim this is _complete_.

adellape replied:
Updating the title to "Container Security Guide".

bfallonf commented on security/index.adoc:
```diff
@@ -0,0 +1,85 @@
+
+This guide provides a high-level walkthrough of the container security measures
+available in {product-title}, from the host layer to the container and
+orchestration layer to the user build and application layer. After reading this
```
The above sentence confused me a little. I think it's the list... Perhaps:

"This guide provides a high-level walkthrough of the container security measures available in {product-title}, including solutions for the host layer, the container and orchestration layer, and the build and application layer."

adellape replied:
:+1:

bfallonf commented on security/index.adoc:
```diff
@@ -0,0 +1,85 @@
+
+Basic concepts covered in this guide include links to further reading provided
+elsewhere in the {product-title} product documentation as well as other Red Hat
+and external documentation.
```
The above probably goes without saying, and perhaps isn't necessary to mention.

adellape replied:
:+1:  Axed. :hocho:

bfallonf commented on security/index.adoc:
```diff
@@ -0,0 +1,85 @@
+and external documentation.
+
+[[security-intro-what-are-containers]]
+== What Are Containers?
```
I like this, and I think it's a much better way to explain containers than what we have anywhere else in the docs. But is it the right place for it? Wouldn't it be easier to put this in the Architecture guide and just reference it here?

Note: I see you link to the Architecture guide in the"Further Reading" links down the bottom, so the reader is being directed there anyway.

bfallonf commented on security/index.adoc:
```diff
@@ -0,0 +1,85 @@
+[[security-intro-enterprise]]
+== Securing Containers for Enterprise Use
+
+However, enterprise use of containers requires strong security. In many ways,
```
Is there a reason this is starting with "However,"? I can't really see what it's linking to. I think if it's a new section it's safe to just jump into the info.

adellape replied:
Removed.

bfallonf commented on security/hosts_multitenancy.adoc:
```diff
@@ -0,0 +1,105 @@
+toc::[]
+
+[[security-hosts-the-os-matters]]
+== The OS Matters
```
I'd spell out Operating System here.

bfallonf commented on security/hosts_multitenancy.adoc:
```diff
@@ -0,0 +1,105 @@
+up each individual container.
+
+To take full advantage of this packaging and deployment technology, an
+operations team needs the right environment for running containers. Operations
```
Is "operations team" something we say? Should it be something like: 

"To take full advantage of this packaging and deployment technology, the cluster administrator must change their operations depending on the operating system (OS)."

?

bfallonf commented on security/hosts_multitenancy.adoc:
```diff
@@ -0,0 +1,105 @@
+operations team needs the right environment for running containers. Operations
+needs an operating system (OS) that can secure containers at the boundaries:
+securing the host kernel from container escapes and securing containers from
+each other.
```
I have no idea what the above is saying. Maybe:

"You need an OS that can secure container so that the host kernel is secure and containers are secure from each other".

Either way, wouldn't it be easier to just say something like "OpenShift is availble on RHEL and Atomic Host, which give X benefits"? I feel like this intro para is telling the reader they can install on anything, "but make sure you can do these secure things, otherwise you'll be in trouble", which I feel is misleading and not correct...

adellape replied:
The point is more about RHEL/Atomic vs other OSes, and that RHEL/Atomic provides XYZ things.

adellape replied:
Updated that sentence to: "You need an
operating system (OS) that can secure the host kernel and secure containers from
each other."

bfallonf commented on security/hosts_multitenancy.adoc:
```diff
@@ -0,0 +1,105 @@
+
+In Linux, containers are really just a special type of process, so your approach
+to securing containers should be the same as your approach to securing any
+running process. Dropping privileges is important and still best practice.
```
s/best practice/recommended

adellape replied:
:+1:

bfallonf commented on security/hosts_multitenancy.adoc:
```diff
@@ -0,0 +1,105 @@
+Containers should run as a non-root user. There are multiple levels of security
+and a lot of history involved in securing containers on Linux.
+
+Linux namespaces, SELinux, CGroups and Secure Computing Mode (seccomp) are four
```
Oxford comma after CGroups

adellape replied:
Embarrassing... This sentence has been axed at this point though.

bfallonf commented on security/hosts_multitenancy.adoc:
```diff
@@ -0,0 +1,105 @@
+to securing containers should be the same as your approach to securing any
+running process. Dropping privileges is important and still best practice.
+Containers should run as a non-root user. There are multiple levels of security
+and a lot of history involved in securing containers on Linux.
```
What does "a lot of history" mean?

adellape replied:
Phrase has been removed by this point.

bfallonf commented on security/hosts_multitenancy.adoc:
```diff
@@ -0,0 +1,105 @@
+
+- _CGroups_ (or control groups) limit, account for, and isolate the resource usage
+(CPU, memory, disk I/O, network, etc.) of a collection of processes. CGroups are
+used to ensure your container will not be stomped on by another container on the
```
What does "stomped on" mean?

adellape replied:
Changed to "impacted".

bfallonf commented on security/hosts_multitenancy.adoc:
```diff
@@ -0,0 +1,105 @@
+
+As a proof point of the security features available to you with RHEL and RHEL
+Atomic Host, RHEL 7.1 recently received Common Criteria certification, including
+certification of the Linux Container Framework Support.
```
It's almost like the above should have to exclamation points...

I'd suggest this be changed to something like "RHEL 7.1 also comes with X security options." and if you're still keen on keeping this para in here, I'd include a link to the features that got RHEL this certification.

sferich888 replied:
Making call outs like this will lead customers to ask, about XYZ certifications, FIPS or PCI are two big ones. In wonder if statements about these need to be had here as well.

bfallonf commented on security/container_content.adoc:
```diff
@@ -0,0 +1,53 @@
+== What's In Your Container?
+
+When it comes to security, what is inside your container matters. For some time
+now, applications and infrastructures have been composed from readily available
```
s/For some time no, applications and infrastructures have been composed from/Applications and infrastructures are composed of

adellape replied:
:+1:

bfallonf commented on security/container_content.adoc:
```diff
@@ -0,0 +1,53 @@
+When it comes to security, what is inside your container matters. For some time
+now, applications and infrastructures have been composed from readily available
+components, many of which are open source packages such as the Linux operating
+system, JBoss Web Server, PostgreSQL and Node.js.
```
Comma after PostgreSQL (and in the line above, before the "such as")

adellape replied:
Oy.

bfallonf commented on security/container_content.adoc:
```diff
@@ -0,0 +1,53 @@
+Containerized versions of these packages are now also readily available, so that
+you do not have to build your own. But, as with any code you download from some
+external source, you need to know where they originally came from, who built
+them and whether there is any malicious code inside them. You need to know:
```
The "You need to know" doens't match the list. I'd suggest "Ask yourself the following:" Or change the list to "If the contents of the container will compromise your infrastructure" "If there are any vulnerabilities in the application layer" and "If the runtime and operating system (OS) layers are up to date".

adellape replied:
Changed to just "Ask yourself".

bfallonf commented on security/container_content.adoc:
```diff
@@ -0,0 +1,53 @@
+- Are the runtime and OS layers up to date?
+
+[[security-content-scanning]]
+== Scanning Container Content
```
Can the two sections here be condensed to one (If so, I'd suggest keeping the "Scanning Container Content" title)? They talking about the same thing, and the second even begins with "To help answer these questions..."

bfallonf commented on security/registries.adoc:
```diff
@@ -0,0 +1,92 @@
+containers_. Immutable containers are containers that will never be changed
+while running.
+
+Container images are a nicely packaged, layered collection of binaries. When a
```
"nicely" is an adverb, which are advised against.

adellape replied:
Gone.

bfallonf commented on security/registries.adoc:
```diff
@@ -0,0 +1,92 @@
+[[security-registries-where-do-your-containers-come-from]]
+== Where Do Your Containers Come From?
+
+As discussed, there are tools you can use to scan and track the contents of your
```
So this is referencing the "Scanning Container Content" topic? Should it link to it? Or even be included with that content?

sferich888 replied:
+1

bfallonf commented on security/registries.adoc:
```diff
@@ -0,0 +1,92 @@
+Container images are a nicely packaged, layered collection of binaries. When a
+bug is discovered in a deployed binary, you do not edit the binary in place; you
+fix the bug in the source code, rebuild the binary, and replace the broken
+binary with the newly rebuilt binary.
```
There's a few "you do not X"'s here. Are these supposed to be recommendations? Like "don't do X". I'm not actually sure what they're saying...

bfallonf commented:
@adellape I stopped reading about halfway, because, sorry, I went a little crazy with comments.

I think this is very marketing-speak-heavy, and I think there's a lot of contextual/architectural/reference information here that is outlined in other parts of the docs (though, granted, it does a good job of explaining the point), but that means it's not particularly in the correct place. All that made me think that maybe I went a little too deep too soon, so I stopped, cos I'm not how through this you are on a scale of 1->10.

Let me know if I'm wrong about the comments I've made, or if you want a deeper read-through.

aweiteka commented:
Great start. I agree this is a bit too high level for what the title suggests. HOWEVER, getting this published soon at this detail level is much preferred over coordinating deeper dives. We need to encourage SME's to contrib an appropriate level of detail.

One topic I missed is a discussion on secrets. Doesn't have to be in the first draft but it's an important platform feature. Could just be an intro with pointers to other parts of the doc.

adellape commented:
@ahardin-rh @bfallonf Thanks for notes, made some changes. Updated preview:

http://file.rdu.redhat.com/~adellape/032217/security_guide/security/

@aweiteka A few brief mentions on secrets added here:

http://file.rdu.redhat.com/~adellape/032217/security_guide/security/build_process.html#security-build-securing-inputs-during-builds

http://file.rdu.redhat.com/~adellape/032217/security_guide/security/deployment.html#security-deployment-secrets-configmaps

Also added a section about platform cert management:

http://file.rdu.redhat.com/~adellape/032217/security_guide/security/securing_container_platform.html#security-platform-cert-mgmt

adellape commented:
Merging this to get a first publish in Origin docs only for now.

sferich888 commented on security/hosts_multitenancy.adoc:
```diff
@@ -0,0 +1,109 @@
+To take full advantage of this packaging and deployment technology, an
+operations team needs the right environment for running containers. You need an
+operating system (OS) that can secure the host kernel and secure containers from
+each other.
```
I wonder if parts of http://rhelblog.redhat.com/2015/07/29/architecting-containers-part-1-user-space-vs-kernel-space/ should be re-used or referenced here?

sferich888 commented on security/hosts_multitenancy.adoc:
```diff
@@ -0,0 +1,109 @@
+the attack surface on the physical server. That said, security must still be
+monitored: one guest VM may be able to use hypervisor bugs to gain access to
+another VM or the host kernel. And, when the OS needs patching, it must be
+patched on all guest VMs using that OS.
```
We need to think about how https://github.com/kubevirt may change the way we message this, or double down on the messaging for areas such as this (Security).

sferich888 commented on security/index.adoc:
```diff
@@ -0,0 +1,88 @@
+containers is a lot like securing any running process. You need to think about
+security throughout the layers of the solution stack before you deploy and run
+your containers. You also need to think about security throughout the
+application and container life cycle.
```
Linking to RHEL security guides here, and showing / proving that the same security guides for RHEL apply here may help better make this clear to customers.

sferich888 commented on security/network_defense.adoc:
```diff
@@ -0,0 +1,78 @@
+consider network defense.
+
+A typical public cloud container service, such as Google Container Engine (GKE),
+Azure Container Services, and AWS Container Service, are single tenant services.
```
In the future, this may not be true (multi-tennant deployments) may be more common.

sferich888 commented on security/network_defense.adoc:
```diff
@@ -0,0 +1,78 @@
+They let you run your containers on the virtual machine (VM) cluster that you
+spin up. {product-title} allows you to take a single Kubernetes cluster and
+segment the traffic to make multi-tenant clusters the isolate different users,
+teams, applications and environments within a single cluster.
```
How does this help security? What your describing here or with this section seem to be a feature pitch, for betting using your resources. We may just need to expand on this more to discuss how (or how the following sections) are going to help you secure your infrastructure / applications.

sferich888 commented on security/network_defense.adoc:
```diff
@@ -0,0 +1,78 @@
+namespaces, each collection of containers (each pod) gets its own entire IP and
+port range to bind to, thereby isolating pod networks from each other on the
+node. Pods from different projects cannot send packets to or receive packets
+from pods and services of a different project. You can use these features to
```
We need to call out when using our multi-tennant plugin (its now the default, but was not with some versions of OpenShift). An then highlight that further or better granularity to this is offered by other plugins.

sferich888 commented on security/rbac.adoc:
```diff
@@ -0,0 +1,24 @@
+
+Roles and Access Management in a Cluster Federation
+
+In July of 2016, Kubernetes 1.3 introduced Kubernetes Federated Clusters.  Federation is useful for very large scale, global deployments. Multiple clusters can be useful for high-availability across multiple availability zones or to enable common management of deployments or migrations across multiple cloud providers, such as AWS, Google Cloud, and Azure.
```
Not sure I like this here, as https://github.com/openshift/origin/  states: 

Feature | Stage (version) | Description
------------ | ------------- |  -------------
Federated Clusters | Alpha (1.3), Beta (1.4, 1.5),	Tech Preview (1.3, 1.4, 1.5) | A Kubernetes federation server may be used against Origin clusters with the appropriate credentials today. Known issues include tenant support in federation and the ability to have consistent access control between federation and normal clusters. **No Origin specific binary is being distributed for federation at this time.**

sferich888 commented on security/registries.adoc:
```diff
@@ -0,0 +1,104 @@
+
+Consuming security updates is particularly important when managing _immutable
+containers_. Immutable containers are containers that will never be changed
+while running.
```
Good place to introduce atomic scan

sferich888 commented on security/registries.adoc:
```diff
@@ -0,0 +1,104 @@
+
+- More on immutable containers in {product-title}:
+** _{product-title} Architecture_: xref:{sourcedir}architecture/core_concepts/builds_and_image_streams.adoc#image-streams[Image Streams]
+** _{product-title} Developer Guide_: xref:{sourcedir}dev_guide/managing_images.adoc#referencing-images-in-image-streams[Referencing Images in Image Streams]
```
We may want to link to https://access.redhat.com/blogs/product-security/posts/container-security-scoring

sferich888 commented on security/securing_container_platform.adoc:
```diff
@@ -0,0 +1,164 @@
+- GitHub / GitLab
+- Google
+- OpenID Connect
+- and more.
```
This content seem to be duplicated.

sferich888 commented on security/storage.adoc:
```diff
@@ -0,0 +1,84 @@
+
+And, of course, you should take advantage of the security features available in
+your chosen storage solution.
+
```
Should we talk about encryption of disks at all?

sferich888 commented on security/container_content.adoc:
```diff
@@ -0,0 +1,64 @@
+- xref:{sourcedir}using_images/index.adoc#using-images-index[_{product-title} Using Images_]
+** Reference documentation on framework, database, and service container images
+provided by Red Hat for use on {product-title}
+
```
Should we also link to https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux_atomic_host/7/html/container_security_guide/

adellape commented:
@sferich888 Planning to further address your comments here after https://github.com/openshift/openshift-docs/issues/4432 merges.
